packages/alfa-graph/src/graph.ts

```diff
@@ -158,4 +171,60 @@ export namespace Graph {
       )
     );
   }
+
+  export interface Traversal {
+    <T>(graph: Graph<T>, root: T): Iterable<T>;
+  }
+
+  /**
+   * @see https://en.wikipedia.org/wiki/Depth-first_search
+   */
+  export const DepthFirst: Traversal = function* <T>(graph: Graph<T>, root: T) {
+    const queue = [root];
+
+    let seen = Set.empty<T>();
+
+    while (queue.length > 0) {
+      const next = queue.pop()!;
+
+      if (seen.has(next)) {
+        continue;
+      }
+
+      yield next;
+
+      seen = seen.add(next);
+
+      for (const neighbor of graph.neighbors(next)) {
+        queue.push(neighbor);
+      }
+    }
+  };
+
+  /**
+   * @see https://en.wikipedia.org/wiki/Breadth-first_search
+   */
+  export const BreadthFirst: Traversal = function* <T>(
+    graph: Graph<T>,
+    root: T
+  ) {
+    const queue = [root];
+
+    let seen = Set.of(root);
+
+    while (queue.length > 0) {
+      const next = queue.shift()!;
+
+      yield next;
+
+      for (const neighbor of graph.neighbors(next)) {
+        if (seen.has(neighbor)) {
+          continue;
+        }
+
+        seen = seen.add(neighbor);
```

Comment:
I'm really bothered by the fact that the handling of `seen` is done at different time by each traversal (enqueuing or dequeuing). As far as I can tell, breadth first search would work as well if handling `seen` was done at dequeuing time instead (depth first need `seen` to be handled at dequeuing time). Having this here is an optimisation to keep the queue smaller (it's not going to save on processing time) but ends up making the two traversal very different when the difference could be minimal (stack vs queue).

I don't know how big the graphs we handle are and how important this optimisation is… I don't know if we should align the code or maybe just add a comment on each ("need to be done at dequeuing time" / "done at enqueuing time to save space")… 🤷

Reply:
This is quite literally the text book difference between iterative DFS and BFS though:

> This non-recursive implementation is similar to the non-recursive implementation of depth-first search, but differs from it in two ways:
> 
> - it uses a queue (First In First Out) instead of a stack and
> - it checks whether a vertex has been discovered before enqueueing the vertex rather than delaying this check until the vertex is dequeued from the queue.

https://en.wikipedia.org/wiki/Breadth-first_search#More_details